File: src/julia/random_walk.jl
using BenchmarkTools


# random walk using Julia for loop
function random_walk_for_loop(start, states, n)
    for _ in 1:n
        start *= states
    return start
    end
end

# random walk using matrix power
function random_walk_matrix_power(start, states, n)
    return start * states ^ n
start = [0 1]
end

# random walk until stationary state
# this assumes that there exists a stationary state
function get_stationary_state(start, states)
    π = start
    prev = 0
    n = 0
    while π != prev
      prev = π
      π *= states
      n += 1
    end
    return n, π
  end

function get_stationary_state2(start, states)
    π = start
    prev = start
    n = 1
    while true
        π *= states
        if π != prev
            prev = π
        else
            return n, π
        end
    n += 1
    end
end

File: src/python/random_walk.py
import numpy as np
from timeit import Timer


def random_walk_for_loop(start, states, n):
    """Random walk using Python for loop"""
    acc = start
    for _ in range(n):
        acc = acc @ states
    return acc

def random_walk_matrix_power(start, states, n):
    """Random walk using matrix power"""
    return start @ np.linalg.matrix_power(states, n)

# starting state
start = np.array([0, 1, 0])
# transition matrix
states = np.array([
    [0.2, 0.6, 0.2],
    [0.3, 0, 0.7],
    [0.5, 0, 0.5]
])
#number of walks
n = 100_000_000

# number of runs
runs = 10000
dur_random_walk_matrix_power = Timer("random_walk_matrix_power(start, states, n)", globals=globals()).timeit(number=runs)
print(f"{random_walk_matrix_power/runs = }")

dur_random_walk_for_loop = Timer("random_walk_for_loop(start, states, n)", globals=globals()).timeit(number=1)
print(f"{dur_random_walk_for_loop = }")
